nixos/syncthing: declare ignore patterns (resolves #268282)

zorrobert opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This PR adds an option to declaratively add ignore patterns to a syncthing folder. Patterns are added as a list of strings which is send to the syncthing service via the rest api. 

This resolves #268282.

## Things done

I've been using this for a few months now and it works with no issues.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

itepastra commented:
I think you need to use [one of these](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) for the issue to close correctly when this is merged

Masrkai commented:
this should have been there from the start with syncthing it's a basic feature please merge it already

zorrobert commented:
Hi, sorry to bother you, but I wanted to ask for someone to review or merge this PR. If you have questions I'll try my best to answer them and make the changes needed to get this merged.
If you don't have the time to review/merge right now that's fine, I just wanted to ask for an update since the PR has been open for some time now.
@ck3d @Jayman2000 @h33p @jfly @bjornfor

ck3d reviewed (approved):
LGTM

jfly commented on nixos/modules/services/networking/syncthing.nix:
```diff
@@ -209,6 +211,13 @@ let
                 ''
                   ${injectSecretsJqCmd} ${jsonPreSecretsFile} | curl --json @- -X POST ${s.baseAddress}
                 ''
+                /*
+                  Check if we are configuring a folder which has ignore patterns.
+                  If it does, write the ignore patterns to the rest API.
+                */
+                + lib.optionalString ((conf_type == "dirs") && (new_cfg.ignorePatterns != [ ])) ''
+                  curl -d '{"ignore": ${builtins.toJSON new_cfg.ignorePatterns}}' -X POST ${s.ignoreAddress}?folder=${new_cfg.id}
+                ''
```
What happens if I configure some ignore patterns, and then remove the code that configures the ignore patterns? IIUC, won't I still have the ignore patterns configured?

zorrobert replied:
I just noticed that using `ignorePatterns = [];` as default value causes some issues, I'm going to change it to be null instead. That way, if the option is undefined (null) the current ignore patterns won't be overridden. If the option is defined, the user hopefully read the option description and is aware that his manually set ignore pattern will be overridden.
Also, now you can use `ignorePatterns = [];` if you want your pattern to be empty, previously you had to use `ignorePatterns = [ "// comment as filler" ];` to get around the `ignorePatterns != [ ]` check you highlighted.
Thank you for the review, I wouldn't have found this issue otherwise and the change makes the option much more intuitive.

jfly replied:
That definitely sounds like an improvement!

zorrobert committed:
nixos/syncthing: add option for ignore patterns

Syncthing can be configured to ignore certain files and
directories via ignore patterns. This PR adds a new nix option to
set these patterns declaratively.

doronbehar reviewed (approved):
Verified the tests pass with this change, and that docs look good.
